### PR TITLE
Use express instead of native http module

### LINKED ISSUE #472: use express instead of http in oauth2 guide
Moved into own issue:

> I think the guide should use express instead of HTTP native library
cause the native library is very complicated for newbie/example, and express is easier than native library.
Just suggestion, don't mind me.
Server-side: https://pastebin.com/RvtZBTZ3 (edit to not expired bin)
HTML: https://pastebin.com/mZwYQwi3 (edit to not expired bin)

_Originally posted by @KurokuTetsuya in https://github.com/discordjs/guide/issues/466#issuecomment-683608408_

### PR DESCRIPTION
**Pull request by KurokuTetsuya**
**Please describe the changes this PR makes and why it should be merged:**
 - Change nodejs native http module to express

closes #472 
Correct me if i wrong. thanks

### CODE REVIEW

**Comment by tipakA** [comment]
Related to #472, i assume

**Comment by KurokuTetsuya** [comment]
Yes, these changes are related to #472

**Comment by advaith1** [comment]
if you put exactly "closes #472" or "resolves #472" in the description then github will automatically close the issue when this or is merged

File: code-samples/oauth/simple-oauth-webserver/index.html [hunk 1]
```diff
@@ -20,18 +20,17 @@
...
 				if (stateParameter !== atob(decodeURIComponent(urlState))) {
 					return console.log('You may have been clickjacked!');
 				}
 
-				const accessToken = fragment.get("access_token");
-				const tokenType = fragment.get("token_type");
+				const accessToken = fragment.access_token;
+				const tokenType = fragment.token_type;
 
 				fetch('https://discordapp.com/api/users/@me', {
```
Comment (Fyko) [inline code-samples/oauth/simple-oauth-webserver/index.html:35]:
```suggestion
				fetch('https://discord.com/api/users/@me', {
```

File: code-samples/oauth/simple-oauth-webserver/index.js [hunk 2]
```diff
@@ -1,58 +1,43 @@
...
+	    const data = {
+	        client_id: 'client_id',
+	        client_secret: 'client_secret',
+	        grant_type: 'authorization_code',
+	        redirect_uri: 'http://localhost:3000/',
+	        code: accessCode,
+	        scope: 'identify',
+	    };
+
+	    const res = await fetch('https://discordapp.com/api/oauth2/token', {
```
Comment (Fyko) [inline code-samples/oauth/simple-oauth-webserver/index.js:19]:
```suggestion
	    const res = await fetch('https://discord.com/api/oauth2/token', {
```

File: code-samples/oauth/simple-oauth-webserver/index.js [hunk 3]
```diff
@@ -1,58 +1,43 @@
...
-	}
-
-	if (urlObj.pathname === '/') {
-		responseCode = 200;
-		content = fs.readFileSync('./index.html');
+		});
+		const info = await res.json();
+		token_type = info.token_type;
+		access_token = info.access_token;
+		const post = await fetch('https://discordapp.com/api/users/@me', {
```
Comment (Fyko) [inline code-samples/oauth/simple-oauth-webserver/index.js:29]:
```suggestion
		const post = await fetch('https://discord.com/api/users/@me', {
```

File: guide/oauth2/README.md [hunk 4]
```diff
@@ -89,11 +89,11 @@ You can see that by clicking `Authorize`, you are allowing the application to ac
...
+			const fragment = parseQuery(window.location.href);
 
-			if (fragment.has("access_token")) {
-				const accessToken = fragment.get("access_token");
-				const tokenType = fragment.get("token_type");
+			if (fragment.access_token) {
+				const accessToken = fragment.access_token;
+				const tokenType = fragment.token_type;
 
 				fetch('https://discordapp.com/api/users/@me', {
```
Comment (Fyko) [inline guide/oauth2/README.md:98]:
```suggestion
				fetch('https://discord.com/api/users/@me', {
```

File: guide/oauth2/README.md [hunk 5]
```diff
@@ -208,15 +221,20 @@ const data = {
 	scope: 'the scopes',
 };
 
-fetch('https://discordapp.com/api/oauth2/token', {
+const res = await fetch('https://discordapp.com/api/oauth2/token', {
```
Comment (Fyko) [inline guide/oauth2/README.md:224]:
```suggestion
const res = await fetch('https://discord.com/api/oauth2/token', {
```

File: guide/oauth2/README.md [hunk 6]
```diff
@@ -234,19 +252,15 @@ Now try visiting your OAuth2 url and authorizing your application. Once you're r
...
-	body: data,
-})
-	.then(res => res.json())
-	.then(info => fetch('https://discordapp.com/api/users/@me', {
-		headers: {
-			authorization: `${info.token_type} ${info.access_token}`,
-		},
-	}))
-	.then(console.log);
+const userPost = await fetch('https://discordapp.com/api/users/@me', {
```
Comment (Fyko) [inline guide/oauth2/README.md:257]:
```suggestion
const userPost = await fetch('https://discord.com/api/users/@me', {
```

**Review by Fyko — CHANGES_REQUESTED** [review]
Domain changes

**Review by almostSouji — CHANGES_REQUESTED** [review]
required: https://github.com/discordjs/guide/pull/485#pullrequestreview-500526637

**Comment by KurokuTetsuya** [comment]
Gonna fix ci soon

File: code-samples/oauth/simple-oauth-webserver/index.html [hunk 7]
```diff
@@ -41,20 +40,28 @@
 					.then(res => res.json())
 					.then(response => {
 						console.log(response);
-						const { username, discriminator } = response;
-						document.getElementById('info').innerText += ` ${username}#${discriminator}`;
+                        const { username, discriminator } = response;
+                        console.log(username, discriminator);
```
Comment (Danktuary) [inline code-samples/oauth/simple-oauth-webserver/index.html:44]: No need for the log here. Indentation is also off.

File: guide/oauth2/README.md [hunk 8]
```diff
@@ -6,30 +6,30 @@ OAuth2 enables application developers to build applications that utilize authent
...
-http.createServer((req, res) => {
-	let responseCode = 404;
-	let content = '404 Error';
-
-	if (req.url === '/') {
-		responseCode = 200;
-		content = fs.readFileSync('./index.html');
-	}
+app.get('/', (request, response) => {
+	response.status(200).send('index.html', { root: '.' });
```
Comment (Danktuary) [inline guide/oauth2/README.md:24]:
```suggestion
	response.status(200).sendFile('index.html', { root: '.' });
```

File: guide/oauth2/README.md [hunk 9]
```diff
@@ -6,30 +6,30 @@ OAuth2 enables application developers to build applications that utilize authent
...
+});
 
-	res.writeHead(responseCode, {
-		'content-type': 'text/html;charset=utf-8',
-	});
+// 404 Route, Keep this route as the last routes
```
Comment (Danktuary) [inline guide/oauth2/README.md:27]:
```suggestion
// Catch-all route; keep this as the last route
```

File: guide/oauth2/README.md [hunk 10]
```diff
@@ -208,15 +221,20 @@ const data = {
...
 	method: 'POST',
 	body: new URLSearchParams(data),
 	headers: {
 		'Content-Type': 'application/x-www-form-urlencoded',
 	},
-})
-	.then(res => res.json())
-	.then(console.log);
+});
+const info = await res.json();
```
Comment (Danktuary) [inline guide/oauth2/README.md:231]:
```suggestion
const { token_type: tokenType, access_token: accessToken } = await res.json();
```

Update the relevant code as well.

File: guide/oauth2/README.md [hunk 11]
```diff
@@ -170,35 +179,39 @@ What you did in the quick example was go through the `implicit grant` flow, whic
...
 	console.log(`The access code is: ${accessCode}`);
 }
 
-if (urlObj.pathname === '/') {
-	responseCode = 200;
-	content = fs.readFileSync('./index.html');
-}
+if (token_type && access_token && request.query.state) {
+	response.status(200).redirect(`/?code=${request.query.code}&state=${request.query.state}&access_token=${access_token}&token_type=${token_type}`);
+} else { response.sendFile('index.html', { root: '.' }); }
```
Comment (Danktuary) [inline guide/oauth2/README.md:204]: Either make this an early return or place the `else {}` content on its own line.

File: guide/oauth2/README.md [hunk 12]
```diff
@@ -170,35 +179,39 @@ What you did in the quick example was go through the `implicit grant` flow, whic
...
 ```
 
-Now you have to exchange this code with Discord for an access token. To do this, you need your `client_id` and `client_secret`. If you've forgotten them, head over to [your applications](https://discord.com/developers/applications) and get them. You can use `node-fetch` to make requests to Discord; you can install it with `npm i node-fetch`.
+Now you have to exchange this code with Discord for an access token. To do this, you need your `client_id` and `client_secret`. If you've forgotten them, head over to [your applications](https://discordapp.com/developers/applications) and get them. You can use `node-fetch` to make requests to Discord; you can install it with `npm i node-fetch`. Place this code after `request.query.code` statement.
```
Comment (Danktuary) [inline guide/oauth2/README.md:207]:
```suggestion
Now you have to exchange this code with Discord for an access token. To do this, you need your `client_id` and `client_secret`. If you've forgotten them, head over to [your applications](https://discord.com/developers/applications) and get them. You can use `node-fetch` to make requests to Discord; you can install it with `npm i node-fetch`. Place this code after your `if (request.query.code)` statement.
```

File: guide/oauth2/README.md [hunk 13]
```diff
@@ -170,35 +179,39 @@ What you did in the quick example was go through the `implicit grant` flow, whic
...
+	const accessCode = request.query.code;
 	console.log(`The access code is: ${accessCode}`);
 }
 
-if (urlObj.pathname === '/') {
-	responseCode = 200;
-	content = fs.readFileSync('./index.html');
-}
+if (token_type && access_token && request.query.state) {
+	response.status(200).redirect(`/?code=${request.query.code}&state=${request.query.state}&access_token=${access_token}&token_type=${token_type}`);
```
Comment (Danktuary) [inline guide/oauth2/README.md:203]: This if statement should be moved inside the `if (request.query.code)` statement, because it utilizes the code part of the URI, as well as other data from the `await res.json()` call. You also use `request.query.code` here instead of the `accessCode` variable created above.

File: guide/oauth2/README.md [hunk 14]
```diff
@@ -170,35 +179,39 @@ What you did in the quick example was go through the `implicit grant` flow, whic
 
 #### Authorization code grant
 
-Unlike the quick example, you need an OAuth2 url where the `response_type` is `code`. Once you've obtained it, try visiting the link and authorizing your application. You should notice that instead of a hash, the redirect url now has a single query parameter appended to it like `?code=ACCESS_CODE`. Modify your `index.js` file to pull the parameter out of the url if it exists. You can use the `url` module to do this for us.
+Unlike the quick example, you need an OAuth2 url where the `response_type` is `code`. Once you've obtained it, try visiting the link and authorizing your application. You should notice that instead of a hash, the redirect url now has a single query parameter appended to it like `?code=ACCESS_CODE`. Modify your `index.js` (on `/` route) file to pull the parameter out of the url if it exists.
```
Comment (Danktuary) [inline guide/oauth2/README.md:182]:
```suggestion
Unlike the quick example, you need an OAuth2 url where the `response_type` is `code`. Once you've obtained it, try visiting the link and authorizing your application. You should notice that instead of a hash, the redirect url now has a single query parameter appended to it like `?code=ACCESS_CODE`. Modify your `index.js` (on the `/` route) file to pull the parameter out of the url if it exists.
```

File: code-samples/oauth/simple-oauth-webserver/index.js [hunk 15]
```diff
@@ -1,58 +1,43 @@
...
-			}))
-			.then(userRes => userRes.json())
-			.then(console.log);
-	}
-
-	if (urlObj.pathname === '/') {
-		responseCode = 200;
-		content = fs.readFileSync('./index.html');
+		});
+		const info = await res.json();
```
Comment (Danktuary) [inline code-samples/oauth/simple-oauth-webserver/index.js:26]:
```suggestion
		const { token_type: tokenType, access_token: accessToken } = await res.json();
```
As well as update the 2 lines below this and the 2 `let` declarations at the top of the function.

File: code-samples/oauth/simple-oauth-webserver/index.js [hunk 16]
```diff
@@ -1,58 +1,43 @@
-const http = require('http');
-const fs = require('fs');
-const url = require('url');
+const express = require('express');
+const app = express();
 const fetch = require('node-fetch');
 
-const port = 53134;
```
Comment (Danktuary) [inline code-samples/oauth/simple-oauth-webserver/index.js:6]: This shouldn't be removed.

File: code-samples/oauth/simple-oauth-webserver/index.js [hunk 17]
```diff
@@ -1,58 +1,43 @@
...
+	if (token_type && access_token && request.query.state) {
+		response.status(200).redirect(`/?code=${request.query.code}&state=${request.query.state}&access_token=${access_token}&token_type=${token_type}`);
+	} else { response.sendFile('index.html', { root: '.' }); }
+});
 
-	res.write(content);
-	res.end();
-})
-	.listen(port);
+app.listen(3000);
```
Comment (Danktuary) [inline code-samples/oauth/simple-oauth-webserver/index.js:43]: This should be `app.listen(port)` after restoring the `port` variable.

File: code-samples/oauth/simple-oauth-webserver/index.html [hunk 18]
```diff
@@ -20,18 +20,17 @@
 			return randStr;
 		}
 
-		window.onload = () => {
-			const fragment = new URLSearchParams(window.location.hash.slice(1));
-
-			if (fragment.has("access_token")) {
-				const urlState = fragment.get("state");
+		window.onload = function () {
+            const fragment = parseQuery(window.location.href);
```
Comment (Danktuary) [inline code-samples/oauth/simple-oauth-webserver/index.html:24]: The `parseQuery` function doesn't seem to work as intended. Here are some quick tests I did:


Changing it from a `new URLSearchParams()` fragment variable would also require you to rephrase a few sentences in the paragraphs on the guide page. I think it's better to remove the custom function and keep it as it was.

**Review by Danktuary — CHANGES_REQUESTED** [review]
There are still a few areas to fix the indentation on as well.

File: code-samples/oauth/simple-oauth-webserver/index.html [hunk 19]
```diff
@@ -41,20 +40,28 @@
...
 			}
-		}
+        }
+        function parseQuery(queryString) {
+            var query = {};
+            var pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
+            for (var i = 0; i < pairs.length; i++) {
+                var pair = pairs[i].split('=');
+                query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || '');
+            }
```
Comment (NotSugden) [inline code-samples/oauth/simple-oauth-webserver/index.html:62]:
```suggestion
            const query = {};
            const pairs = (queryString.startsWith('?') ? queryString.slice(1) : queryString).split('&');
            for (const param of pairs) {
                const [name, value] = param.split('=');
                query[decodeURIComponent(name)] = decodeURIComponent(value || '');
            }
```

File: code-samples/oauth/simple-oauth-webserver/index.html [hunk 20]
```diff
@@ -20,18 +20,17 @@
 			return randStr;
 		}
 
-		window.onload = () => {
-			const fragment = new URLSearchParams(window.location.hash.slice(1));
-
-			if (fragment.has("access_token")) {
-				const urlState = fragment.get("state");
+		window.onload = function () {
```
Comment (NotSugden) [inline code-samples/oauth/simple-oauth-webserver/index.html:23]:
```suggestion
		window.onload = () => {
```
why was this changed to `function`?

File: guide/oauth2/README.md [hunk 21]
```diff
@@ -112,6 +112,15 @@ You can see that by clicking `Authorize`, you are allowing the application to ac
...
 		}
+		function parseQuery(queryString) {
+            var query = {};
+            var pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
+            for (var i = 0; i < pairs.length; i++) {
+                var pair = pairs[i].split('=');
+                query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || '');
+            }
+            return query;
+        }
```
Comment (NotSugden) [inline guide/oauth2/README.md:123]: same as previous suggestion in the other example

**Review by NotSugden — CHANGES_REQUESTED** [review]

File: guide/oauth2/README.md (hunk 13 above)
Comment (Danktuary) [reply guide/oauth2/README.md:203]: Actually, I don't think this redirect is needed. You need the access token/token type/state query params in the index.html file, but only when using the the `response_type=token` param used in the index.html code at first. When moving it over to the server side, it shows how to implement the OAuth code and switches the `response_type` to `code`.  At the bottom of the page it also mentions:
> To maintain security, store the access token server side but associate it with a session ID that you generate for the user.

**Comment by WiseDevHelper** [comment]
@Danktuary do you think I can open a new PR and complete what was left out?
The previous contributor was too busy to finish this one though, it's almost done.

**Comment by Danktuary** [comment]
@WiseDevHelper Yeah that'd be fine, go ahead.
